src/xtgeo/grid3d/_grid_import_xtgcpgeom.py

```diff
@@ -128,41 +128,71 @@ def import_hdf5_cpgeom(self, mfile, ijkrange=None, zerobased=False):
         meta = json.loads(jmeta, object_pairs_hook=OrderedDict)
 
         req = meta["_required_"]
+        ncol = req["ncol"]
+        nrow = req["nrow"]
+        nlay = req["nlay"]
 
         if ijkrange is not None:
-            incoord, inzcorn, inactnum, ncol2, nrow2, nlay2 = _partial_read(
+            incoord, inzcorn, inactnum, ncol, nrow, nlay = _partial_read(
                 h5h, req, ijkrange, zerobased
             )
+            if (
+                "subgrids" in meta["_required_"]
+                and meta["_required_"]["subgrids"] is not None
+            ):
+                meta["_required_"]["subgrids"] = filter_subgrids_partial(
+                    meta["_required_"]["subgrids"], *ijkrange[-2:], nlay, zerobased
+                )
         else:
             incoord = grp["coord"][:, :, :]
             inzcorn = grp["zcorn"][:, :, :, :]
             inactnum = grp["actnum"][:, :, :]
 
-    for myattr in reqattrs:
-        if "subgrid" in myattr:
-            self.set_subgrids(reqattrs["subgrids"])
-        else:
-            setattr(self, "_" + myattr, req[myattr])
-
-    if ijkrange:
-        self._ncol = ncol2
-        self._nrow = nrow2
-        self._nlay = nlay2
-
     self._coordsv = incoord.astype("float64")
     self._zcornsv = inzcorn.astype("float32")
     self._actnumsv = inactnum.astype("float32")
 
-    if self._xshift != 0.0 or self._yshift != 0.0 or self._zshift != 0.0:
-        self._coordsv[:, :, 0::3] += self._xshift
-        self._coordsv[:, :, 1::3] += self._yshift
-        self._coordsv[:, :, 2::3] += self._zshift
-        self._zcornsv += self._zshift
-        self._xshift = 0.0
-        self._yshift = 0.0
-        self._zshift = 0.0
+    handle_metadata(self, meta, ncol, nrow, nlay)
 
-    self._metadata.required = self
+
+def filter_subgrids_partial(subgrids, k1, k2, nlay, zerobased):
+    """
+    Filters away subgrids that are not part of the partial grid
+
+    Args:
+        subgrids: The OrderedDict of subgrids.
+        k1: Start of subgrid layers (can be "min" to mean 0 or 1 dependent on zerobased)
+        k2: End of subgrid layers (cna be "max" to mean nlay or nlay -1
+            dependent on zerobased.
+        nlay: Original number of layers
+        zerobased: Whether indexing starts with 0 or 1.
+
+    Returns:
+        New Orderedict which is the subgrids dictionary with out of range
+        subgrids removed or truncated.
+
+    """
+    if k1 == "min":
+        k1 = 0 if zerobased else 1
+
+    if k2 == "max":
+        k2 = nlay - 1 if zerobased else nlay
+
+    if not zerobased:
+        k1 -= 1
+        k2 -= 1
```

Comment:
If `k1 == "min"` is True and zerobased, `k1` is set to 0 and not changed - that's fine I suppose. If it is not zerobased, `k1` is first set to 1 and then reduced to 0 - should `not zerobased` have the index set to 0?

Reply:
This copies the logic from `_partial_read` . Before line 181, k1 and k2 are either zero based or one based. After line 183 they are converted to zero based.

Reply:
Added a comment with some explanation.